[wasm] Make runtime_is_initialized promise callbacks one-shot

lambdageek opened the pull request:
Throw if `runtime_is_initialized_resolve` or `runtime_is_initialized_reject` is called more than once

lambdageek committed:
[wasm] Make runtime_is_initialized promise callbacks one-shot

Throw if runtime_is_initialized_resolve or
runtime_is_initialized_reject is called more than once

kg commented:
You don't need to do it, but it's worth considering whether we should also explicitly throw if resolve() is called after reject(), or vice-versa. Looking at the spec it appears that the normal behavior for mistakes like this (and for double-reject) is silently ignoring the call, and we want to throw.

![image](https://user-images.githubusercontent.com/198130/173534139-b7d0224c-2289-47d7-a18e-a404ef392eae.png)

Also, according to documentation if the executor (the function being passed resolve/reject) throws it will reject the promise, so that would make calling reject() after that an error. Thankfully in this case it's not possible for our executor to throw, but it's worth knowing if you use this pattern in the future :)

lambdageek commented:
The original motivation for this PR is that I was confused by the error handling in `startup.ts` and was pretty convinced that we would call `reject` multiple times.  Multiple-`accept` isn't really even a suspicion on any control flow path AFAICT.

I guess it might be worthwhile to also guard against `resolve`-after-`reject` (and vice versa, though it seems unlikely).

> You don't need to do it, but it's worth considering whether we should also explicitly throw if resolve() is called after reject(), or vice-versa. 
> …
> Also, according to documentation if the executor (the function being passed resolve/reject) throws it will reject the promise, so that would make calling reject() after that an error. Thankfully in this case it's not possible for our executor to throw, but it's worth knowing if you use this pattern in the future :)

I think there's probably value in making a `GuardedPromise<T>` (or something), but I'm not sure if dotnet is the place for it.

pavelsavara commented:
What would be the problem with calling reject multiple times, even if that happened ? It would reject the promise only once, right ? Do we need to learn about the second time ?

lambdageek committed:
Add a slightly generalized GuardedPromise<T> object

Protects against multiple-resolve, multiple-reject, reject after resolve and
 resolve after reject.

 Does not protect against the executor throwing.

lambdageek commented on src/mono/wasm/runtime/guarded-promise.ts:
```diff
@@ -0,0 +1,31 @@
+/// A Promise<T> that guards against multiple-resolve, multiple-reject, reject-after-accept and accept-after-reject.
```
```suggestion
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

/// A Promise<T> that guards against multiple-resolve, multiple-reject, reject-after-accept and accept-after-reject.
```

lambdageek committed:
add header

pavelsavara commented:
> After line 345 we should really return instead of falling through to `bindings_lazy_init`

AFAIK `wasm_exit` would throw or really exit in nodejs.

pavelsavara commented:
> > > After line 345 we should really return instead of falling through to `bindings_lazy_init`
> > 
> > 
> > AFAIK `wasm_exit` would throw or really exit in nodejs.
> 
> I'm talking about the browser case - we reject the promise, skip the `if (ENVIRONMENT_IS_SHELL)` and fall through to `bindings_lazy_init`.

I see, I guess we should throw explicitly there.
